src/client/dfuse/ops/read.c

```diff
@@ -12,159 +12,160 @@
+	     position >= oh->doh_ie->ie_end_off || position + len <= oh->doh_ie->ie_start_off)) {
 		off_t pos_ra = position + len + READAHEAD_SIZE;
 
 		DFUSE_TRA_DEBUG(oh, "Returning zeros");
-		skip_read = true;
+		mock_read = true;
 
 		if (pos_ra <= oh->doh_ie->ie_stat.st_size &&
-		    ((oh->doh_ie->ie_start_off == 0 &&
-				oh->doh_ie->ie_end_off == 0) ||
-				(position >= oh->doh_ie->ie_end_off ||
-					pos_ra <= oh->doh_ie->ie_start_off))) {
+		    ((oh->doh_ie->ie_start_off == 0 && oh->doh_ie->ie_end_off == 0) ||
+		     (position >= oh->doh_ie->ie_end_off || pos_ra <= oh->doh_ie->ie_start_off))) {
 			readahead = true;
 		}
-	} else if (oh->doh_caching &&
-		len < (1024 * 1024) &&
-		oh->doh_ie->ie_stat.st_size > (1024 * 1024)) {
+	} else if (oh->doh_caching && len < (1024 * 1024) &&
+		   oh->doh_ie->ie_stat.st_size > (1024 * 1024)) {
 		/* Only do readahead if the requested size is less than 1Mb and
 		 * the file size is > 1Mb
 		 */
 
 		readahead = true;
 	}
 
-	if (readahead) {
-		if (!fs_handle->dpi_info->di_wb_cache)
-			buff_len += READAHEAD_SIZE;
-	} else {
-		if (!skip_read) {
-			rc = daos_event_init(&ev->de_ev,
-					     fs_handle->dpi_eq, NULL);
-			if (rc != -DER_SUCCESS)
-				D_GOTO(err, rc = daos_der2errno(rc));
-
-			ev->de_req = req;
-			ev->de_complete_cb = dfuse_cb_read_complete;
-			async = true;
-		}
-	}
+	if (fs_handle->dpi_info->di_wb_cache)
+		readahead = false;
+
+	if (readahead)
+		buff_len += READAHEAD_SIZE;
 
 	D_ALLOC(buff, buff_len);
 	if (!buff)
 		D_GOTO(err, rc = ENOMEM);
 
-	ev->de_sgl.sg_nr = 1;
 	d_iov_set(&ev->de_iov, (void *)buff, buff_len);
-	ev->de_sgl.sg_iovs = &ev->de_iov;
-
-	if (skip_read) {
+	ev->de_sgl.sg_iovs  = &ev->de_iov;
+	ev->de_oh           = oh;
+	ev->de_req          = req;
+	ev->de_req_len      = len;
+	ev->de_req_position = position;
+	ev->de_sgl.sg_nr    = 1;
+
+	if (mock_read) {
 		ev->de_len = buff_len;
-	} else {
-		rc = dfs_read(oh->doh_dfs, oh->doh_obj, &ev->de_sgl, position,
-			      &ev->de_len, async ? &ev->de_ev : NULL);
-		if (rc != -DER_SUCCESS) {
-			DFUSE_REPLY_ERR_RAW(oh, req, rc);
-			D_FREE(buff);
-			D_FREE(ev);
-			return;
-		}
-	}
-
-	if (async) {
-		/* Send a message to the async thread to wake it up and poll
-		 * for events
-		 */
-		sem_post(&fs_handle->dpi_sem);
+		dfuse_cb_read_complete(ev);
+		D_FREE(ev);
 		return;
 	}
 
-	if (ev->de_len <= len) {
-		if (ev->de_len == 0)
-			DFUSE_TRA_DEBUG(oh, "Truncated read, %#zx-%#zx (EOF)",
-					position, position);
-		else if (ev->de_len != len)
-			DFUSE_TRA_DEBUG(oh, "Truncated read, %#zx-%#zx",
-					position, position + ev->de_len - 1);
-		DFUSE_REPLY_BUF(oh, req, buff, ev->de_len);
+	rc = daos_event_init(&ev->de_ev, fs_handle->dpi_eq, NULL);
+	if (rc != -DER_SUCCESS)
+		D_GOTO(err_buff, rc = daos_der2errno(rc));
+
+	ev->de_complete_cb = dfuse_cb_read_complete;
+
+	rc = dfs_read(oh->doh_dfs, oh->doh_obj, &ev->de_sgl, position, &ev->de_len, &ev->de_ev);
+	if (rc != -DER_SUCCESS) {
```

Comment:
Since dfs returns system error numbers, do we prefer this?
```suggestion
	rc = dfs_read(oh->doh_dfs, oh->doh_obj, &ev->de_sgl, position, &ev->de_len, &ev->de_ev);
	if (rc != 0) {
```

Reply:
Yes, that's a mistake on my part, will fix.